pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,817 @@
+	selectorItemKey string
+}
+
+// selectorItem represents an individual label selector. It's the actual object that will be matched with label sets.
+// Groups having same label selector will share a selectorItem.
+type selectorItem struct {
+	// The label selector that will be used for matching.
+	selector *types.GroupSelector
+	// The keys of the groupItems that share the selectorItem.
+	groupItemKeys sets.String
+	// The keys of the labelItems that match the selectorItem.
+	labelItemKeys sets.String
+}
+
+var _ Interface = &GroupEntityIndex{}
+
+// GroupEntityIndex implements Interface.
+//
+// It abstracts label set and label selector from entities and groups and does actual matching against the formers to
+// avoid redundant calculation given that most entities (Pod or ExternalEntity) actually share labels. For example, Pods
+// that managed by a deployment controller have same labels.
+//
+// It maintains indexes between label set and label selector so that querying label sets that match a label selector and
+// reversed queries can be performed with a constant time complexity. Indirectly, querying entities that match a group
+// and reversed queries can be performed with same complexity.
+//
+// The relationship of the four items are like below:
+// entityItem <===> labelItem <===> selectorItem <===> groupItem
+type GroupEntityIndex struct {
+	lock sync.RWMutex
+
+	// entityItems stores all entityItems.
+	entityItems map[string]*entityItem
+
+	// labelItems stores all labelItems.
+	labelItems map[string]*labelItem
+	// labelItemIndex is nested map from entityType to Namespace to keys of labelItems.
+	// It's used to filter potential labelItems when matching a Namespace scoped selectorItem.
+	labelItemIndex map[entityType]map[string]sets.String
+
+	// groupItems stores all groupItems.
+	groupItems map[string]*groupItem
+
+	// selectorItems stores all selectorItems.
+	selectorItems map[string]*selectorItem
+	// selectorItemIndex is nested map from entityType to Namespace to keys of selectorItems.
+	// It's used to filter potential selectorItems when matching an labelItem.
+	// Cluster scoped selectorItems are stored under empty Namespace "".
+	selectorItemIndex map[entityType]map[string]sets.String
+
+	// namespaceLabels stores label sets of all Namespaces.
+	namespaceLabels map[string]labels.Set
+
+	// eventHandlers is a map from group type to a list of handlers. When a type of group's updated, the corresponding
+	// event handlers will be called with the group name provided.
+	eventHandlers map[GroupType][]eventHandler
+
+	// eventChan is channel used for calling eventHandlers asynchronously.
+	eventChan chan string
+
+	// Numbers used to determine whether the GroupEntityIndex has been initialized with the full lists of Pods,
+	// Namespaces, and ExternalEntities.
+	// If a kind's max count is greater than or equal to its initial count, it's considered synced.
+	maxPodCount     int
+	curPodCount     int
+	initialPodCount *int
```

Comment:
Probably add some comments to explain nil means initialPodCount is not initalized yet.
What you think using -1 to indicate initialPodCount is not set?

Reply:
done